File: preview/gsCompilarEjecutarNET/OpcionesForm.vb
Public Class OpcionesForm

    Public CargarUltimo As Boolean
    Public ColorearAlCargar As Boolean

    Public Sub New(colFics As List(Of String))

        ' Esta llamada es exigida por el diseñador.
        InitializeComponent()

        ' Agregue cualquier inicialización después de la llamada a InitializeComponent().
        Me.ColFics.AddRange(colFics)

    End Sub

    Public Property ColFics As New List(Of String)

    Private Sub OpcionesForm_Load(sender As Object, e As EventArgs) Handles Me.Load
        chkCargarUltimo.Checked = CargarUltimo
        chkColorearCargar.Checked = ColorearAlCargar
        AñadirEventos()
        AsignarItems()

    End Sub

    Private Sub OpcionesForm_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing

        If e.CloseReason = CloseReason.UserClosing Then
            Me.DialogResult = DialogResult.Cancel
        End If
    End Sub

    Private Sub AsignarItems()
        lstFics.Items.Clear()
        For i = 0 To ColFics.Count - 1
            lstFics.Items.Add(ColFics(i))
        Next
    End Sub

    Private Sub AñadirEventos()
        AddHandler lstFics.KeyDown, AddressOf lstFics_KeyDown
        AddHandler btnCancelar.Click, Sub() Me.DialogResult = DialogResult.Cancel
        AddHandler btnAceptar.Click, Sub()
                                         Me.CargarUltimo = chkCargarUltimo.Checked
                                         Me.ColorearAlCargar = chkColorearCargar.Checked
                                         Me.DialogResult = DialogResult.OK
                                     End Sub
        AddHandler btnQuitar.Click, Sub() QuitarSeleccionados()
        AddHandler btnOrdenar.Click, Sub()
                                         ColFics.Sort()
                                         AsignarItems()
                                     End Sub
        AddHandler lstFics.SelectedIndexChanged, Sub() txtFic.Text = If(lstFics.SelectedItem Is Nothing, txtFic.Text, lstFics.SelectedItem.ToString)

    End Sub

    Private Sub lstFics_KeyDown(sender As Object, e As KeyEventArgs)
        If e.KeyCode = Keys.Delete Then
            e.Handled = True
            QuitarSeleccionados()
        End If
    End Sub

    Private Sub QuitarSeleccionados()
        If lstFics.SelectedIndices.Count = 0 Then Return
        For i = lstFics.SelectedIndices.Count - 1 To 0 Step -1
            lstFics.Items.RemoveAt(lstFics.SelectedIndices(i))
        Next
        ColFics.Clear()
        For i = 0 To lstFics.Items.Count - 1
            ColFics.Add(lstFics.Items(i).ToString)
        Next
    End Sub

End Class

File: preview/gsCompilarEjecutarNET cs/OpcionesForm.cs
/*
'------------------------------------------------------------------------------
' Opciones de la aplicación de compilar y ejecutar
'
'
' (c) Guillermo (elGuille) Som, 2020
'------------------------------------------------------------------------------
*/
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace gsCompilarEjecutarNET_cs
{
    public partial class OpcionesForm : Form
    {
        public OpcionesForm()
        {
            InitializeComponent();
        }

        public bool CargarUltimo;
        public bool ColorearAlCargar;

        public OpcionesForm(List<string> colFics)
        {

            // Esta llamada es exigida por el diseñador.
            InitializeComponent();

            // Agregue cualquier inicialización después de la llamada a InitializeComponent().
            this.ColFics.AddRange(colFics);

            this.FormClosing += new System.Windows.Forms.FormClosingEventHandler(this.OpcionesForm_FormClosing);
            this.Load += new System.EventHandler(this.OpcionesForm_Load);
        }

        public List<string> ColFics { get; set; } = new List<string>();

        private void OpcionesForm_Load(object sender, EventArgs e)
        {
            chkCargarUltimo.Checked = CargarUltimo;
            chkColorearCargar.Checked = ColorearAlCargar;
            AñadirEventos();
            AsignarItems();
        }

        private void OpcionesForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (e.CloseReason == CloseReason.UserClosing)
                this.DialogResult = DialogResult.Cancel;
        }

        private void AsignarItems()
        {
            lstFics.Items.Clear();
            for (var i = 0; i <= ColFics.Count - 1; i++)
                lstFics.Items.Add(ColFics[i]);
        }

        private void AñadirEventos()
        {
            lstFics.KeyDown += lstFics_KeyDown;
            btnCancelar.Click += (object s, EventArgs e) => this.DialogResult = DialogResult.Cancel;

            btnAceptar.Click += (object s, EventArgs e) =>
                                    {
                                        this.CargarUltimo = chkCargarUltimo.Checked;
                                        this.ColorearAlCargar = chkColorearCargar.Checked;
                                        this.DialogResult = DialogResult.OK;
                                    };
            btnQuitar.Click += (object s, EventArgs e) => QuitarSeleccionados();
            btnOrdenar.Click += (object s, EventArgs e) =>
                                    {
                                        ColFics.Sort();
                                        AsignarItems();
                                    };
            lstFics.SelectedIndexChanged += (object s, EventArgs e) => 
                                                txtFic.Text = (lstFics.SelectedItem == null 
                                                                ? txtFic.Text 
                                                                : lstFics.SelectedItem.ToString()
                                                               );
        }

        private void lstFics_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Delete)
            {
                e.Handled = true;
                QuitarSeleccionados();
            }
        }

        private void QuitarSeleccionados()
        {
            if (lstFics.SelectedIndices.Count == 0)
                return;

            for (var i = lstFics.SelectedIndices.Count - 1; i >= 0; i += -1)
                lstFics.Items.RemoveAt(lstFics.SelectedIndices[i]);

            ColFics.Clear();

            for (var i = 0; i <= lstFics.Items.Count - 1; i++)
                ColFics.Add(lstFics.Items[i].ToString());
        }
    }
}
